### PR TITLE
Add blog

### PR DESCRIPTION
**Pull request by arthuro555**
Begin of blog. More or less usable, still lacks some functionality and need to be restyled completely (I am very bad at styling :/)

### CODE REVIEW

**Comment by 4ian** [comment]
That's fine we can mix/compare and take the best of both :) This looks like a good start!

Here is a preview of yours: https://5f2436a9dbb0250007b90966--gdevelop-website.netlify.app/

I recommend:
* For styling, look at components already existing. Try to reuse some. You can make new components, feel free to take example of existing blogs to get some inspiration. **Stay simple**, a few simple margins, text size and spacing goes a long way.
* I can see that the index page of the blog is only showing one article - is this normal?
  image

**Comment by arthuro555** [comment]
> I can see that the index page of the blog is only showing one article - is this normal?  

Yes, there is a "hidden" attribute in the two testing articles.

Thanks for the styling tips!

**Comment by 4ian** [comment]
Excellent then! :) 

I'm also not a designer but usually the best thing to do is to "benchmark", which means here going at alternative websites that you know/find and see how they styled their blog, for example a randomly selected: [Itch.io blog](https://itch.io/blog/163598/games-of-the-month-cellular-harvest-hylics-2-and-more), [material-ui blog](https://material-ui.com/blog/material-ui-v4-is-out/), [Roblox](https://blog.roblox.com/2020/07/roblox-physics-building-better-sleep-system/)...

Here I would heavily favor something *very clean and simple*. No need to go fancy for a first version with buttons to share (though that would be cool and not too complex), tags, etc.. A nicely shown `BigTitle` or `Title`, the content centered (like it's done for the rest of the website using `TransparentContainer`), intermediate titles with a smaller `Title`, the date in italics, maybe a different styling for the intro, 2 `BigGhostButton` at the bottom to go the the previous/next blog post, ensure the link/styling of markdown is good aaannnd... that's it? :)

**Comment by arthuro555** [comment]
Ok I think I have an ok styling. Do you see anything that can be improved?
Now I just want to add 2 more things:
1. Blog post thumbnail
2. Credit the author of the post

**Comment by arthuro555** [comment]
The thubnails could be probably done better using sharp but I don't understand how it works well :/
I made a deployement of the version with the new styles here: https://gdblog-preview.netlify.app/blog/

**Comment by Bouh** [comment]
I suggest this what you think ?
image
*nitpicking*

**Comment by 4ian** [comment]
On mobile we should probably remove the margin between the blog article
content and the side of the page, keeping only a small padding/margin :)

I'm fine with titles centered or not, no strong feelings here!

**Comment by 4ian** [comment]
Do you think we could dump entirely the "card" background when reading a blog article? I have a feeling it's not entirely needed.

**Comment by arthuro555** [comment]
I thought it looked good and more in the style of the website (the background would not be good with text directly on it so it let see the background with the style of the rest of the website). Maybe I can ask on the forum or discord what people like the most ? What matters in the end is if the ones who will read it appreciate it after all 😊.

De : Florian Rival<mailto:notifications@github.com>
Envoyé le :dimanche 2 août 2020 19:07
À : 4ian/GDevelop-website<mailto:GDevelop-website@noreply.github.com>
Cc : Arthur Pacaud<mailto:arthur.pacaud@hotmail.fr>; Author<mailto:author@noreply.github.com>
Objet :Re: [4ian/GDevelop-website] [WIP] Add blog (#126)


Do you think we could dump entirely the "card" background when reading a blog article? I have a feeling it's not entirely needed.

**Comment by Bouh** [comment]
Take the containers inclined as background.
I mean the components, `TransparentContainer` `SkewedBorderContainer` `SkewedBackgroundContainer`
Mockup
image

An article can be simply the same page but with only one container.
This fit more to the actual design :)

**Comment by 4ian** [comment]
> Mockup

I like it :) Something along these lines would fit a bit more maybe with the rest of the website :) 

> An article can be simply the same page but with only one container.
This fit more to the actual design :)

Yeah, I think it makes sense!

**Comment by arthuro555** [comment]
🤔 it would fit more with the rest of the website but imo it doesn't look like a blog. It looks really like just idk how to tell it raw information thrown to you. I think the whole "big text" design of the website is good as it is for presenting and is like a PowerPoint. But a blog article is more a long text and needs to be a bit more compact I think

Then it is just my opinion, if you think it's better that way I'll just do with it

**Comment by 4ian** [comment]
It might be fine to go with a "MediumTitle" then if it's really too big.
Also sure a "blog" might reasonate with "article" which might make you think of a paper-like card design, but it's nowhere really mandatory and this metaphor seems a bit obselete now in design - I would not be too concern about it as long as we're delimiting the blog article properly (and as long as we're reducing titles a bit - like you said it might not be adapted otherwise to showing a lot of text :))

**Comment by arthuro555** [comment]
> Also sure a "blog" might reasonate with "article" which might make you think of a paper-like card design

No, I just wanted to "fill the space". I first just dropped everything in a transparent container but it looked really bad and empty without anything in the background, so I used the `SkewedBorderContainer`, but the colors of the text directly on the backround didn't always look good, so I came up with that card to make an empty background behind the text without having nothing and a feeling of an empty page.

For the list of cards I used them because they looked like a good way of showing that info when looking at the list of components from bulma css:
image
When I saw that I just thought "That may look good for GDevelop's blog" so I used it and showed the others on discord and they seemed to appreciate:
image
So I went with it

**Comment by arthuro555** [comment]
After asking on the discord people seem to prefer the card design over the website design, aka this:
image
Over this: 
image

**Comment by 4ian** [comment]
Ok no strong feelings on my side, let's go with the first one if you prefer. Can we remove the skewed background? It's a bit weird to have it on top, it's usually used in the middle of a page to emphasize/separate a section :) 
image

Instead, I recommend making a new container that you can name `GreyContainer` or `BlogContainer` that will be used to give the page a different background than the rest of the website, which is all white :) Useful to distinguish a bit the "traditional" website from articles :)

File: i18n-helpers.js [hunk 1]
```diff
@@ -1,68 +1,68 @@
...
+    }))
+    .map(locale => ({
+      ...locale,
+      path: locale.langCode,
+      name: getLocaleName(locale.langCode),
+      translationRatio: computeTranslationRatio(locale),
+    }));
+
+module.exports = {
+  getAllLocales,
```
Comment (4ian) [inline i18n-helpers.js:67]: Can you revert the changes in these files? Might be because Unix line endings where changed to Windows line endings.

File: i18n/catalog.json [hunk 2]
```diff
@@ -41,6 +41,7 @@
   "Features": "Features",
   "Games": "Games",
   "Game studios & education": "Game studios & education",
+  "Blog": "Blog",
```
Comment (4ian) [inline i18n/catalog.json:44]: Same, revert all the changes in i18n folder, they should not be in this PR.

File: src/components/BlogCard.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      <div className="card">
+        {this.props.title && (
+          <header className="card-header">
+            <p className="card-header-title">{this.props.title}</p>
+            {this.props.date && (
+              <p className="card-header-icon">{this.props.date}</p>
+            )}
+          </header>
+        )}
+        {this.props.thumbnail && (
```
Comment (4ian) [inline src/components/BlogCard.js:16]: Oh I know see that you've been using bulma. Unfortunately we rather try in the website to not use custom style - it's only used because it "resets" the default style to some sane values.
Instead, we use styled-components: don't think in terms of CSS classes, but think in terms of components that are getting some styles. Then your page is just assembling components. 

See how BigTitle, BigGhostButton, TransparentContainer etc.. are done.

It's a slightly different logic - and long to write than bulma, we don't really want to rely on bulma.

File: src/templates/BlogPost.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            <BannerContainer>
+              <NavBarSpacer />
+            </BannerContainer>
+            <Spacer height="10px" style={{backgroundColor: "#f1f1f1"}}/>
+            <BoxContainer
+              title={post.frontmatter.title}
+              subtitle={post.frontmatter.date}
+            >
+              <div dangerouslySetInnerHTML={{ __html: post.html }} />
+              <nav>
```
Comment (4ian) [inline src/templates/BlogPost.js:48]: Make a component for this

File: src/components/BlogCard.js (hunk 3 above)
Comment (arthuro555) [reply src/components/BlogCard.js:16]: > we don't really want to rely on bulma

Why is it there if we can't use it? It confuses me. And that would mean all my work until now was for nothing :/ Not sure I am good enough with css to make something near as good as right now using bulma.

File: i18n-helpers.js (hunk 1 above)
Comment (arthuro555) [reply i18n-helpers.js:67]: I tried but nothing seems to work. I even tried to reset them from master git doesn't detect any change. That's really weird.

**Comment by arthuro555** [comment]
Ok so from the "query side" everything is now done. The only thing left to do is changing the design from using Bulma to styled components. I am really bad at CSS so I would appreciate if someone else (maybe @Bouh 👀) could try to do it.

Maybe @4ian can already review the "query/backend" part of the code already?

**Comment by Bouh** [comment]
I do design and CSS, I put the styles in the components.
I aim to finish tonight.
Then can I push my code here ?

**Comment by Bouh** [comment]
For authors instead of defining everything in an object, it would be better to read the github username in the markdown file and use the github API for avatars and get the real names.
That way we don't need the " blog/authors/index.js" file.

Can you remove the name of the images/gif in the description ?
image

**Comment by arthuro555** [comment]
> Can you remove the name of the images/gif in the description ? 

The text shown is the alt text. I guess you can just not enter any alt text in the markdown to remove it.

> For authors instead of defining everything in an object, it would be better to read the github username in the markdown file and use the github API for avatars and get the real names.
> That way we don't need the " blog/authors/index.js" file.

That was the point of that file, not having to rely on an external service. From how I see the blog it is not reserved to GitHub contributors (hence the anonym User) but if it has to yes we can do that (especially since github has a graphql API).

File: src/components/Blog/Author.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import styled from 'styled-components';
+import { Link } from 'gatsby';
+import getAuthorData from '../../../blog/authors';
+import Avatar from './Avatar';
+
+const Container = styled.div`
+  float: left;
```
Comment (4ian) [inline src/components/Blog/Author.js:8]: Try to use flexbox please :) `float: left` is very old CSS and unreliable. Instead, use `display: flex` in the parent (the container), with `align-items`/`justify-content`.
It should be the parent that decides about the alignment of its children - *not* the children that should decide where they "float".

**Comment by 4ian** [comment]
> That was the point of that file, not having to rely on an external service. From how I see the blog it is not reserved to GitHub contributors (hence the anonym User) but if it has to yes we can do that (especially since github has a graphql API).

Let's not complicate things, I think the time to just enter a name and username is 100x faster and easier than doing a GitHub query, getting the results, checking why it failed, handling the failures in netlify because somehow GitHub was done at the same time we were deploying etc etc etc ;)
Here a simple list will beat anything else more automated as long as we're <50 regular blog contributors - which will be the case for the next year or so ;)

File: gatsby-browser.js [hunk 6]
```diff
@@ -0,0 +1 @@
+require('./static/prism-theme.css');
```
[the remark this answers is not in the document]
Comment (arthuro555) [reply gatsby-browser.js:1]: Idk on the prism documentatiuon it said it had to be there but maybe it doesn't really has to.

File: blog/posts/test-post/index.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+---
+title: Hello World
+date: "07-31-2020 12:12"
+description: "Praesent lacinia, lectus eu rutrum facilisis, orci nisl porta erat, id fringilla augue enim vel ligula. Ut elementum et massa at convallis."
+author: "arthuro555"
+hidden: false
```
[the remark this answers is not in the document]
Comment (arthuro555) [reply blog/posts/test-post/index.md:6]: Yes, Bouh unhid them in his commits while testing

**Comment by arthuro555** [comment]
https://www.netlifycms.org/
This could be potentially interresting for the blog

**Comment by 4ian** [comment]
It could be interesting, but I'm also very happy with markdown files, as it's easy to write by hand :) We can always plug something like this on top of it later, but for now I think we're in a good state with our blog + markdown files.

**Comment by arthuro555** [comment]
Exactly, this cms is made for static files and markdown.

File: src/components/Containers/BoxContainer.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    <div style={{ backgroundColor: '#f1f1f1' }}>
+      <CenteredRow>
+        <div className="tile is-parent" style={{ textAlign: 'initial' }}>
+          <article class="tile is-child is-12 box">
+            {props.title && <div className="title">{props.title}</div>}
+            {props.subtitle && <div className="subtitle">{props.subtitle}</div>}
+            <div className="content">{props.children}</div>
+          </article>
+        </div>
+      </CenteredRow>
```
[the remark this answers is not in the document]
Comment (arthuro555) [reply src/components/Containers/BoxContainer.js:15]: I am pretty sure this component is dead code 🤔

**Comment by arthuro555** [comment]
I couldn't replicate the ENOENT issue, I think I fixed the other netlify build errors

**Comment by 4ian** [comment]
Netlify still failing it seems:

```

9:12:46 PM: Executing user command: npm run build
9:12:46 PM: > gdevelop-website@2.0.0 build /opt/build/repo
9:12:46 PM: > gatsby build
9:12:50 PM: success open and validate gatsby-configs - 0.062s
9:12:52 PM: 27 locales loaded: en,ar,ca,de,el,es,fil,fr,hi,hu,id,it,ja,ko,ms,nl,pl,pt-BR,pt,ru,sl,sr-CS,sr,tr,yo,zh,zh-TW
9:12:52 PM: success load plugins - 1.119s
9:12:52 PM: success onPreInit - 0.022s
9:12:52 PM: success delete html and css files from previous builds - 0.015s
9:12:52 PM: success initialize cache - 0.011s
9:12:52 PM: success copy gatsby files - 0.037s
9:12:52 PM: success onPreBootstrap - 0.011s
9:12:52 PM: success createSchemaCustomization - 0.171s
9:12:53 PM: success Checking for changed pages - 0.000s
9:12:53 PM: success source and transform nodes - 0.819s
9:12:53 PM: success building schema - 0.494s
9:12:53 PM: success createPages - 0.030s
9:12:53 PM: success Checking for changed pages - 0.000s
9:12:56 PM: success createPagesStatefully - 2.471s
9:12:56 PM: success update schema - 0.077s
9:12:56 PM: success onPreExtractQueries - 0.001s
9:12:56 PM: success extract queries from components - 0.487s
9:12:56 PM: success write out redirect data - 0.002s
9:12:56 PM: success Build manifest and related icons - 0.001s
9:12:56 PM: success onPostBootstrap - 0.002s
9:12:56 PM: info bootstrap finished - 9.984s
9:12:57 PM: success run static queries - 0.112s - 1/1 8.93/s
9:12:59 PM: success run page queries - 1.996s - 652/652 326.65/s
9:12:59 PM: success write out requires - 0.018s
9:13:43 PM: failed Building production JavaScript and CSS bundles - 44.582s
9:13:43 PM: error Generating JavaScript bundles failed
9:13:43 PM: Can't resolve '../../components/blog/Author' in '/opt/build/repo/src/components/Blog'
9:13:43 PM: If you're trying to use a package make sure that '../../components/blog/Author' is installed. If you're trying to use a local file make sure that the path is correct.
9:13:43 PM: error Generating JavaScript bundles failed
9:13:43 PM: Can't resolve '../../components/blog/Author' in '/opt/build/repo/src/components/Blog'
9:13:43 PM: If you're trying to use a package make sure that '../../components/blog/Author' is installed. If you're trying to use a local file make sure that the path is correct.
9:13:43 PM: error Generating JavaScript bundles failed
9:13:43 PM: Can't resolve '../../components/blog/BlogBigTitle' in '/opt/build/repo/src/components/Blog'
9:13:43 PM: If you're trying to use a package make sure that '../../components/blog/BlogBigTitle' is installed. If you're trying to use a local file make sure that the path is correct.
9:13:43 PM: error Generating JavaScript bundles failed
9:13:43 PM: Can't resolve '../../components/blog/FormatedDate' in '/opt/build/repo/src/components/Blog'
9:13:43 PM: If you're trying to use a package make sure that '../../components/blog/FormatedDate' is installed. If you're trying to use a local file make sure that the path is correct.
9:13:43 PM: error Generating JavaScript bundles failed
9:13:43 PM: Can't resolve '../../components/blog/FormatedDate' in '/opt/build/repo/src/components/Blog'
9:13:43 PM: If you're trying to use a package make sure that '../../components/blog/FormatedDate' is installed. If you're trying to use a local file make sure that the path is correct.
9:13:43 PM: not finished Generating image thumbnails - 46.707s
9:13:44 PM: npm
9:13:44 PM:  ERR! code ELIFECYCLE
9:13:44 PM: npm
9:13:44 PM:  ERR!
9:13:44 PM: errno
9:13:44 PM:  1
9:13:44 PM: npm ERR!
9:13:44 PM:  gdevelop-website@2.0.0 build: `gatsby build`
9:13:44 PM: npm ERR! Exit status 1
9:13:44 PM: npm ERR!
9:13:44 PM: npm ERR! Failed at the gdevelop-website@2.0.0 build script.
9:13:44 PM: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
9:13:44 PM: npm
9:13:44 PM: ERR!
9:13:44 PM:  A complete log of this run can be found in:
9:13:44 PM: npm
9:13:44 PM: ERR!
9:13:44 PM:      /opt/buildhome/.npm/_logs/2020-08-23T20_13_44_398Z-debug.log
9:13:44 PM: Skipping functions preparation step: no functions directory set
```

**Comment by arthuro555** [comment]
I'll capitalize all imports tomorrow I already have shutted down my PC, it's getting late.

**Comment by arthuro555** [comment]
Thanks @Bouh  for fixing the imports for me :)
Just tried to deloy from netlify on my side and it built fine with Bouh's fix: https://gdblog-preview.netlify.app/blog/

**Comment by 4ian** [comment]
I've not forgotten this, thanks for the patience 🙏
I'll make an article about the new beta version and will publish this! :)

**Comment by 4ian** [comment]
I'll finally publish the first article today or tomorrow :)

**Comment by Bouh** [comment]
Cocorico 🐓!
I'm so happy to have this blog.
We'll be able to communicate more easily with the community and those who follow us!

I would like to add that I have submitted the Discord server for certification and the case will be studied.

**Comment by 4ian** [comment]
Thanks a lot @arthuro555 :) I added some styling to the generated markdown html otherwise titles were not styled, also added styling for lists and a few other stuffs. Apart from that great job! 

This will be useful to announce new releases that are important, and we could also make more generic articles about game creation, talk about some interesting games or creators, etc... Feel free to suggest your ideas :)
